Display all markers for events

localjo opened the pull request:
Fixes #407, but is still a work in progress.

localjo commented:
@Benjaki2 This first commit here fixes the stuff I noticed with OpenLayers. We can either leave it as a part of this PR, and merge it when I'm done here, or you can cherry pick it out into another branch. Just let me know what you want to do.

Benjaki2 commented:
@localjo sounds good! Prob best just to leave it in and merge it with the feature. 👍

localjo commented:
Some outstanding things to take care of;

- [ ] Add SVGs for every marker type
- [ ] Add hover styles to markers
- [ ] Improve selection/deselection of events
- [ ] Cross browser testing (especially SVG implementation)
- [ ] Persist marker state on refresh

localjo commented:
I added SVG makers for every event type we'll expect to see. The markers are similar to the icons in the list, but I added a "map pin" shape behind them and made a few minor tweaks to the iceberg and snow event icons. I'm going to need to update the icons in the list too for consistency, and I'll probably convert the list to use inline SVGs instead of CSS backgrounds. Aside from consistency, that will allow us to change the color of the icons with CSS (to create better hover states, for example).

<img width="645" alt="screen shot 2017-08-16 at 5 16 01 pm" src="https://user-images.githubusercontent.com/516388/29385713-bb35b754-82a6-11e7-9658-2c806de1c3e4.png">

localjo commented:
I've been playing around with different configurations of the SVG files, and we're really limited with what we can do with external files. It would be ideal to have a way to move the SVGs inline on build. But I can't really think of an easy way to add that to our build system and still have the icons work in dev. I guess for now, leaving the SVGs as external files is best, and we can come back to the ability to include them inline when we overhaul our build system. Unless you have any ideas how we could do that now @Benjaki2

localjo commented:
I improved marker visibility. I think this looks pretty nice;

<img width="509" alt="screen shot 2017-08-17 at 10 51 19 am" src="https://user-images.githubusercontent.com/516388/29418454-50a9a8de-833a-11e7-963b-85567473a587.png">

I was also thinking about exploring a brighter color scheme. When I was looking at how other maps make their markers more visible, I noticed that most markers are brightly colored. For now, I think the white backgrounds make these work. But it might be worth having another discussion about colors in the future.

localjo commented:
Added filtering based on current view #466 but it's still buggy on refresh and switching between tabs

localjo commented:
29 days and 52 commits later this is ready for cross browser testing and review!

![giphy](https://user-images.githubusercontent.com/516388/30173688-80eb71fc-93c6-11e7-899e-d2505cd5a0d7.gif)

localjo commented:
Feedback from our weekly meeting;

- [ ] Make animations smoother
- [ ] Fix SVGs in IE11
- [ ] Make tab toolbars fixed position
- [ ] Change cursor on marker hover
- [ ] Remove scale transform for selected markers on hover

localjo commented:
- [ ] Add "Read More" text to notification
- [ ] Change footer text to say "Only showing events on map."
- [ ] Fix footer layout when empty, and spacing of text

rboller commented:
@localjo I think the zooming action is way better now and I can't overload it with rapid-fire event selections.  

The one thing that I think needs to be updated is the "Off-map events are hidden" notice.  It's styled the same as a button and seems like you should be able to click it.  Likewise with the "Showing all active events in list" notice.

I'd advocate a styling change and/or update to what is shown.

@minniewong what do you think?

minniewong commented:
@localjo and @rboller - I agree the styling makes it look like a button. I'm not sure we need text saying that it's showing all the events. Also, I think the wording could be a little different - I will think about what this better wording is :)

localjo commented:
@rboller @minniewong The text saying that we're showing all events is there to fill the empty space in the tab's footer when the "Show All" button isn't displayed. Because of the way those tabs are built, I'd have to write some hacky JavaScript to remove the empty space. But, if you think it looks bad as it is, I can do that.

For the wording, I'm trying to choose something short that fits in the available space, because adding more space to that footer also means writing more hacky JavaScript where we don't need to. :-/ That's why I chose "Off-map events are hidden."

I'd rather not add lots of hacky code for simple layout changes, but the bottom line is how well it works for users, so I'm happy to make the tradeoff if you think it's worth it. Let me know what you think. :)

rboller commented:
@localjo thanks for the background.  Hopefully we wouldn't need to make the code too hacky, but usability and clarity usually are the highest priorities.

Let's see what @minniewong drafts for wording and go from there.

Benjaki2 commented:
@localjo can you add a class like `showing-all-events` that is toggled on/off on the events tab div el and do the styling changes in CSS?

Zoom looks much better! I would consider slowing down some of the animations to give more time for the tiles to load.

localjo commented:
New wording and layout complete. :)

<img width="309" alt="screen shot 2017-09-13 at 11 27 17 am" src="https://user-images.githubusercontent.com/516388/30386636-555e7ae6-9878-11e7-9543-afbeeeb3d0ff.png">

Benjaki2 commented on web/js/layers/wv.layers.sidebar.js:
```diff
@@ -241,81 +250,26 @@ wv.layers.sidebar = wv.layers.sidebar || function(models, config) {
-        .hide();
-      $("#layers-add")
-        .show();
+      $('.ui-tabs-nav li.first').addClass("ui-state-active");
```
nice simplification

Benjaki2 commented on web/js/naturalEvents/wv.naturalEvents.markers.js:
```diff
@@ -0,0 +1,175 @@
+  });
+};
+
+var createBoundingBox = function(coordinates){
```
CreateBoundingBox and createPin are now global. Will want to include these inside the main function.

localjo replied:
Good catch!  😬

localjo commented:
Turns out that the way I'm creating markers from SVG elements has terrible performance in IE, and makes the event tab and markers take up to 30 seconds to activate. Not good. So, first I played around with writing more efficient code to generate the SVG elements, but that only reduced the time by 10-15 seconds. @ZachTRice did some research and found out that IE's performance when rendering many SVG elements is notoriously bad.

So I did an experiment to render all SVG icons as background images with CSS instead of inline, and the performance is good! The main trade off is that we can't control the icon colors with CSS, so hover states will probably just be size changes, without color changes.

localjo commented on buster.js:
```diff
@@ -20,7 +20,7 @@ config["wv"] = {
     "web/ext/main/jscache-gitba01cdc/cache.js",
     "web/ext/proj/proj4js-2.3.3-1/proj4.js",
-    "web/ext/map/openlayers-3.17.1.wv1/ol.js",
+    "node_modules/openlayers/dist/ol.js",
```
@Benjaki2 should this be ol-debug.js?

localjo commented:
@Benjaki2 Can you take a look at how I'm interacting with the model? It seems to work, but I'm not 100% confident that the things I'm doing with the model follow the correct patterns.

Benjaki2 commented:
@localjo Sure thing. Can you point me to any sections of particular concern?

localjo commented:
Sure @Benjaki2 

When I added an event deselection function I had to manually add a trigger to save the change to the model;
https://github.com/nasa-gibs/worldview/pull/463/commits/ce0eb0a4eef208a3363cd12c4908aa6519e1e75d#diff-a40963c60fdce3cf6e87e62c9a5da447R132

I'm not sure if that's the right approach, but it seems to work. However, the selectEvent function doesn't manually trigger a model change, because it's calling `models.date.select` which triggers a change anyway; https://github.com/nasa-gibs/worldview/blob/edf187eff7bc03d7a7f5d1499f18ce4f665ef4e2/web/js/naturalEvents/wv.naturalEvents.ui.js#L104

I don't want to trigger redundant change events that cause the model to save state multiple times when nothing has actually changed, but I'm also not sure if relying on the `models.date.select` trigger will come back to bite us in the future.

Benjaki2 commented:
@localjo found a little css bug. This item in blue should be white. I think 😁  ?
![screen shot 2017-09-18 at 1 09 37 pm](https://user-images.githubusercontent.com/3605988/30554507-c2811784-9c72-11e7-95d4-3138b3c21cc3.png)

Benjaki2 commented on web/js/naturalEvents/wv.naturalEvents.model.js:
```diff
@@ -42,25 +25,29 @@ wv.naturalEvents.model = wv.naturalEvents.model || function(models, config) {
-        wvx.ui.sidebar.selectTab("events");
+    if (!state.e) return;
+    models.wv.events.on('startup', function() {
+      wvx.ui.sidebar.selectTab('events');
```
I would get rid of reference to this global `wvx` object in this file and pass a ui parameter in from `wv.main.js`

Benjaki2 commented:
@localjo I understand your concern with redundant events but after looking at the code it seems that you've set up the triggers in the best way I can think of.

localjo commented on web/index.html:
```diff
@@ -71,7 +71,7 @@
 
                             <link type="text/css" rel="Stylesheet" href="ext/ui/noUiSlider-6.1.10/jquery.nouislider.css">
-                            <link type="text/css" rel="Stylesheet" href="ext/map/openlayers-3.17.1.wv1/ol.css">
+                            <link type="text/css" rel="Stylesheet" href="ext/node_modules/openlayers/dist/ol.css">
```
Oops, I guess I added it by mistake

localjo commented:
Latest changes are deployed to UAT. Going to do some more browser testing tomorrow and make sure that my fixes didn't break anything new.

localjo commented:
Summary of changes in this PR;

* Adds ability to see all event markers at once on the map (Fixes #407)
* Changes event selection function to use event id rather than DOM index
* Improves day selection for Storms and Wildfires
* Makes markers clickable to select events
* Improves event marker style for visibility on the map
* Adds unique SVG map markers for each event type
* Adds hover styles to event markers
* Improves precision of event marker position on map
* Adds map markers for events with a bounding box
* Improves bounding box style
* Moves selected event state into model
* Saves event selection state to URL
* Preserves selected event markers on refresh (Fixes #472)
* Allows linking to selected events
* Adds notification for links to inactive/missing events
* Sorts events in model in reverse chronological order by the event's most recent active date
* Improves animations and fixes animation glitches (Fixes #471)
* Improves order of execution during event selection (layers change after animation)
* Adds code to handle interrupted animations gracefully
* Adds filtering to event list to only show events that are visible on the map (Fixes #466)
* Adds button to show all events in a filtered list
* Adds loading indicator for events panel
* Makes tab footers unique to each tab, and fixed position, rather than calculated
* Preserves active layers when selecting a different event/date of the same type (Fixes #388)
* Improves spacing around buttons in tab footer
* Updates OpenLayers to 4.3.1 and fix broken references
* Renames naturalEvents.map to naturalEvents.markers
* Renames variables and function names for clarity
* Fixes code that removes ignored events/categories from model
* Triggers 'queryResults' event only once
* Removes unused variables and dead code
* Refactors and simplifies events code
* Moves code to correct files for better organization
* Cleans up tab UI
* Adds vector icons for events tab and tab expand/collapse arrow
* Fixes various bugs
* Enhances performance

localjo commented:
And, a more human readable version of the changes for release notes;

New Features

* You can now see all natural events on the map at once. (Fixed #407)
* Selected events are preserved when you refresh the app and you can share links to selected events using the `e` parameter in the URL with the format `?e={event_id},{event_date}`, for example, `?e=EONET_2874,2014-07-24`. (Fixed #472)
* The events list is now automatically filtered to only show events that are visible in the current map view, and a "List All" button has been added to see the full list of active events. (Fixes #466)
* You can now select events by clicking on their marker on the map.
* Events with a bounding box now have a marker in addition to the bounding box.
* Event marker styles have been improved; each event category now has a unique marker. Markers have been given an outline to make them more visible on the map, and they now use cleaner SVG icons instead of raster images.
* Event bounding box styles have been improved to be more visible, cleaner, and more aesthetically pleasing.

Fixes and Minor Changes

* Active layers are now preserved when selecting a different event/date in the same category. (Fixed #388)
* Animations are smoother and some bugs have been worked out. (Fixed #471)
* The event list is now sorted more reliably in reverse chronological order.
* The default date selection for Storms and Wildfires has been improved.
* Event markers are now positioned more precisely on the map.
* A warning notification has been added for links to inactive/missing events.
* A loading indicator has been added to the events panel, since the events may load slowly in some circumstances.
* Spacing around the buttons at the bottom of tabs has been improved.
* OpenLayers was updated to version 4.3.1.
* Vector icons were added for the events tab and the tab expand/collapse arrows.
* Extensive bug fixes, performance enhancements, and refactoring.
